commit c8d674946a687f40a126166edf470733fc8ede02
Author: ruc_zhangxiaohui <553441439@qq.com>
Date:   Thu Feb 4 04:19:08 2021 +0800

    uses the PerceptibleReciprocal() to prevent the divide-by-zero from occurring (#3194)

    Co-authored-by: Zhang Xiaohui <ruc_zhangxiaohui@163.com>
    (cherry picked from commit 329dd528ab79531d884c0ba131e97d43f872ab5d)

    This backports the fix for CVE-2021-20244 to IM6.

diff --git a/magick/visual-effects.c b/magick/visual-effects.c
index b51f0f62d..9f15dea71 100644
--- a/magick/visual-effects.c
+++ b/magick/visual-effects.c
@@ -1103,11 +1103,11 @@ MagickExport Image *ImplodeImage(const Image *image,const double amount,
           */
           factor=1.0;
           if (distance > 0.0)
-            factor=pow(sin((double) (MagickPI*sqrt((double) distance)/
-              radius/2)),-amount);
+            factor=pow(sin((double) (MagickPI*sqrt((double) distance)*
+              PerceptibleReciprocal(radius)/2)),-amount);
           status=InterpolateMagickPixelPacket(image,image_view,
-            UndefinedInterpolatePixel,(double) (factor*delta.x/scale.x+
-            center.x),(double) (factor*delta.y/scale.y+center.y),&pixel,
+            UndefinedInterpolatePixel,(double) (factor*delta.x*PerceptibleReciprocal(scale.x)+
+            center.x),(double) (factor*delta.y*PerceptibleReciprocal(scale.y)+center.y),&pixel,
             exception);
           if (status == MagickFalse)
             break;
